traits/trait_types.py

```diff
@@ -1901,58 +1899,96 @@ def init_fast_validate(self, *args):
+       is ``True``).
 
     Parameters
     ----------
     *args
-        The enumeration of all legal values for the trait.  The expected
-        signatures are either:
-
-        - a collection.  The default value is the first item in the
-          collection. The collection should conform to the
-          collections.abc.Collection interface. That is, it at least
-          provides the __contains__, __len__ and __iter__ methods.
-          Note that although the types str, bytes, and bytearray
-          conform to the collection interface, these are handled
-          as discrete units.
-        - a single default value, combined with the values keyword
-          argument.
-        - a default value, followed by a single list enum.Enum, tuple or
-          collection conforming to collections.abc.Collection
-        - arbitrary positional arguments each giving a valid value.
-    values : str
-        The name of a trait holding the legal values.  A default value may
-        be provided via a positional argument, otherwise the first item in
-        the collection is used as the default value. Note that if the
-        collection does not have a notion of order like a set, the default
-        value will be an arbitrary element from the set.
+        The enumeration of all valid values for the trait. For a static
+        enumeration trait (where the *values* keyword argument is not given)
+        the supported signatures for ``args`` are as follows:
+
+        (collection,)
+            A nonempty collection of valid values. The default is the first
+            element of the collection, in iteration order.
+        (default, collection)
+            The default value, followed by a nonempty collection of valid
+            values. The default should be an element of the collection, but
+            this is not checked.
+        (item1, item2, ..., itemn)
+            One or more items giving the valid values for the collection.
+            The default is *item1*.
+
+        For a dynamic enumeration trait, where the *values* keyword argument
+        is given, the supported signatures for ``args`` are:
+
+        ()
+            No arguments given. In this case the default is the first item
+            of the collection, in iteration order.
+        (default,)
+            The default value for the collection.
+
+        For the static case, the ambiguity in the signatures is resolved
+        as follows: if ``args`` has length ``1`` or ``2``, ``arg[-1]`` can be
+        iterated over, and ``arg[-1]`` is not an instance of ``str``, ``bytes``
+        or ``bytearray``, then ``arg[-1]`` is assumed to give the collection
+        of values. Otherwise, all elements of ``args`` are assumed to be
+        items in the collection. Thus the first two signatures are safe
+        from ambiguity, and it's recommended to use one of these two signatures
+        in preference to the third form.
```

Comment:
Yes, I had it that way before, but the splitting of the allowed signatures seemed worse than the context reset. It reads better to me with the enumeration of the allowed values first, then the following notes.

Reply:
Yep, I can see the tradeoff. It is fine the way it is now: My brain did fail to reset the context (because it failed to read) but then it realised the "third form" does not make sense for the dynamic case and corrected itself. :)